### PR TITLE
refactor of processing uris

### LINKED ISSUE #5790: [Shell] Uri Navigation Spec when using GoToAsync
# Shell.Navigation.GotoAsync

Parent #2415 

## Route structure
`//route (required) / page (optional) ? QueryParams (optional)`

- Route defines the path to a ShellContent element.These are the views that exist as part of the Shell visual hierarchy. 
- Pages don't exist in the visual hierarchy but can be pushed onto the stack from wherever you are. For example an Item Details page won't be defined in the Shell XAML file but you will want to push it onto the stack from different places.
- QueryParams are any parameters you want passed into your destination page

## Route Decomposition

Given the following URI
`//store/category/edit?CategoryId=12`

The routing system will first try to locate what part of the URI represents the `ShellContent`. So if the structure of the shell page is defined as
```
store
--search
--category
about
```

The routing system will first try to locate a full path that matches `//store/category/edit`. Once that's not found it will remove the last segment and try to locate a route at `//store/category`.

This will cause the store page to be visible with the category tab selected. Now it has to locate the edit page you want to use. The routing system will perform the following route searches and once a route is found then whatever page is found there will be set as the content page.

```
//store/category/edit
//store/edit
//edit
```

Once the page is found that is what the visible content will be set to. This will not create a page stack, the stack will only have one item on it.

Currently the pages will be registered in code
```c#
RegisterRoute("//store/category/edit", typeof(EditCategory));
RegisterRoute("//people/search/edit", typeof(EditSearch));
```

#### Incorrect Routes
If a user tries
`//store/category/ediit`

Then it should throw an exception much like you'd see on MVC
"Edit not found at any of the following locations <list of locations searched>"

#### Duplicates Routes
If a duplicate route is registered it should fail with an exception right when the duplicate route is added. If the XAML defined causes a duplicate route then an exception will be thrown on startup with a helpful message

### Route debugging
Add a static property to routes (That's enabled on the templates) which will dump the current set of routes to the debug output. This will dump the routes out when the app first spins up and then when any changes are made to the routes structure it will dump out all available routes

# API

### Properties

| API | Description |
| ------------- | ------------- |
| Navigation | Adds a static property onto shell for accessing Navigation. |

### Properties

| API | Description |
| ------------- | ------------- |
| GoToAsync |  Provide a uri that you want to navigate to |
 
### Properties

| API | Description |
| ------------- | ------------- |
| DebuggingEnabled | Setting this to true will dump information about the route structure to the output window  |

# Scenarios
Given the following setup
```XAML
<Shell>
    <FlyoutItem Route="xamtube">
		<Tab Route="home" />
		<Tab Route="library" />
		<Tab Route="store" />
	</FlyoutItem>
</Shell>
```

```cs
RegisterRoute("LoginPage", typeof(LoginPage));
RegisterRoute("StoreItem", typeof(StoreItem));
RegisterRoute("//xamtube/library/edit", typeof(EditLibraryContentPage));
RegisterRoute("//xamtube/library/store", typeof(EditStoreContentPage));
RegisterRoute("edit", typeof(EditSettings));
```

*Set root of shell to default content page for xamtube*
```C#
GotoAsync("//xamtube")
```

*Set xam tube as root of shell and activate library tab*
```C#
GotoAsync("//xamtube/library") 
```

*User will only see Login Page and no Shell Navigation*
```C#
GotoAsync("//LoginPage") 
```

*Store tab will be selected on the xamtube page. The content will be set to the StoreItem Page.*
```C#
GotoAsync("//xamtube/store/StoreItem?ItemId=123") 
```

*This creates a navigation stack with the home tab and then the store tab. If the user hits the back button it will go to the home page*
```C#
GotoAsync("//xamtube/home") 
GotoAsync("xamtube/store") 
```

*This will set `xamtube/library` as root of shell with `EditLibraryContentPage` as the content*

```C#
GotoAsync("//xamtube/library/edit") 
```

#### Existing INavigation mappings ####
```C#
public interface INavigation
{
	// Contains stack of Pages on shell contents
	IReadOnlyList<Page> NavigationStack { get; }
	
	//can be used to insert a page somewhere on stack. This is the same as pushing a PageContent
	void InsertPageBefore(Page page, Page before);
	
	// locate content in stack that matches this page and remove it from the stack
	void RemovePage(Page page);
	
	// Go back in stack
	Task<Page> PopAsync();
	
	// Go back in stack
	Task<Page> PopAsync(bool animated);
	
	// pop everything except base
	Task PopToRootAsync();
	
	// pop everything except base
	Task PopToRootAsync(bool animated);

	// push a PageContent onto Shell stack
	Task PushAsync(Page page);

	// push a PageContent onto Shell stack
	Task PushAsync(Page page, bool animated);
	
	// push a PageContent onto Shell stack without navigation structure (at some point we will let people make this prettier)
	Task PushModalAsync(Page page);
	
	// push a PageContent onto Shell stack without navigation structure (at some point we will let people make this prettier)
	Task PushModalAsync(Page page, bool animated);
	
	// Contains stack of any modal pages pushed
	IReadOnlyList<Page> ModalStack { get; }
	
	// Close current Modal page
	Task<Page> PopModalAsync();
	
	// Close current Modal page
	Task<Page> PopModalAsync(bool animated);

}
```

## Additional things to flesh out and other specs in the works
- Embedded a stack into the uri?
- Navigating by the x:Name of a Shell<Thing> 
- Overloading GoToAsync with a NavigationBehavior structure to allow for more complex parameters and used to indicate things like transitions

# Difficulty : medium

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###
- Changed it so Uri's are processed before navigation occurs into a NavigationRequest class. This lets me throw more intelligent exceptions when Uri is wrong plus I can get fancier with decomposing the Uri
- Added lots of behavior to just try as hard as I can to match a uri request to something
- added relative matching that walks up the route hiearchy

### Issues Resolved ### 
- fixes #5790

### API Changes ###

Added:

These classes need to still be fleshed out a bit but they are the beginnings of extracting a uri into a request.  

- public class NavigationRequest
- public class RequestDefinition

Changed:
 - IShellItemController.GoToParts to take NavigationRequest
 - IShellSectionController.GoToParts to take NavigationRequest
 

 
 None

### Platforms Affected ### 
- Core/XAML (all platforms)
- iOS
- Android
- UWP

### Behavioral/Visual Changes ###
Review  #5790 to see how uri's are interpreted

### Testing Procedure ###
- Navigate your little heart out
- added lots of UI Tests

### PR Checklist ###

- [x] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Core/Shell/Shell.cs [hunk 1]
```diff
@@ -343,40 +351,159 @@ void IShellController.UpdateCurrentState(ShellNavigationSource source)
...
+			public WhatToDoWithTheStack StackRequest { get; }
+			public string Query { get; }
+			public string Fragment { get; }
+			public RequestDefinition Request { get; }
+		}
+
+		List<RequestDefinition> BuildAllTheRoutes()
+		{
+			List<RequestDefinition> routes = new List<RequestDefinition>();
+			// todo make better maybe
```
Comment (paymicro) [inline Xamarin.Forms.Core/Shell/Shell.cs]: If this function need only for search `RequestDefinition` by FullUri or ShortUri then it maybe better only return this RequestDefinition when finding matches with uri.
```C#
RequestDefinition GetRequestRoute(string FullUri, string ShortUri)
{
... 
if(longUri == FullUri || shortUri == ShortUri)
return new RequestDefinition(longUri, shortUri, item, section, content);
}
```

File: Xamarin.Forms.Core/Shell/Shell.cs [hunk 2]
```diff
@@ -343,40 +351,159 @@ void IShellController.UpdateCurrentState(ShellNavigationSource source)
...
+				for (var j = 0; j < item.Items.Count; j++)
+				{
+					var section = item.Items[j];
+
+					for (var k = 0; k < section.Items.Count; k++)
+					{
+						var content = section.Items[k];
+
+						string longUri = $"{RouteScheme}://{RouteHost}/{Routing.GetRoute(this)}/{Routing.GetRoute(item)}/{Routing.GetRoute(section)}/{Routing.GetRoute(content)}";
+						string shortUri = $"{RouteScheme}://{RouteHost}/{Routing.GetRoutePathIfNotImplicit(this)}{Routing.GetRoutePathIfNotImplicit(item)}{Routing.GetRoutePathIfNotImplicit(section)}{Routing.GetRoutePathIfNotImplicit(content)}";
```
Comment (paymicro) [inline Xamarin.Forms.Core/Shell/Shell.cs]: I'd like to optimize it. =)

File: Xamarin.Forms.Core/Shell/BaseShellItem.cs [hunk 3]
```diff
@@ -1,9 +1,11 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using Xamarin.Forms.Internals;
 
 namespace Xamarin.Forms
 {
+	[DebuggerDisplay("Title = {Title}, Route = {Route}")]
```
Comment (paymicro) [inline Xamarin.Forms.Core/Shell/BaseShellItem.cs]: 💯

File: Xamarin.Forms.Core.UnitTests/ShellUriHandlerTests.cs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+			Assert.AreEqual(new Uri("app://host/shellroute/path"), ShellUriHandler.ConvertToStandardFormat(shell, CreateUri("app:/path")));
+
+			Assert.AreEqual(new Uri("app://host/shellroute/path"), ShellUriHandler.ConvertToStandardFormat(shell, CreateUri("app://host/path")));
+			Assert.AreEqual(new Uri("app://host/shellroute/path"), ShellUriHandler.ConvertToStandardFormat(shell, CreateUri("app:/host/path")));
+
+			Assert.AreEqual(new Uri("app://host/shellroute/path"), ShellUriHandler.ConvertToStandardFormat(shell, CreateUri("app://shellroute/path")));
+			Assert.AreEqual(new Uri("app://host/shellroute/path"), ShellUriHandler.ConvertToStandardFormat(shell, CreateUri("app:/shellroute/path")));
+
+			Assert.AreEqual(new Uri("app://host/shellroute/path"), ShellUriHandler.ConvertToStandardFormat(shell, CreateUri("app://host/shellroute/path")));
+			Assert.AreEqual(new Uri("app://host/shellroute/path"), ShellUriHandler.ConvertToStandardFormat(shell, CreateUri("app:/host/shellroute/path")));
```
Comment (paymicro) [inline Xamarin.Forms.Core.UnitTests/ShellUriHandlerTests.cs]: can make variable `new Uri("app://host/shellroute/path")` ?
Also need add `CreateUri` with backslash separator `CreateUri("\\\\path"))` and `CreateUri("\\path"))`

File: Xamarin.Forms.Core/Routing.cs [hunk 5]
```diff
@@ -11,35 +11,55 @@ public static class Routing
...
 		{
-			if (source.StartsWith(ImplicitPrefix, StringComparison.Ordinal))
+			if (IsImplicit(source))
 				return source;
-			return ImplicitPrefix + source;
+			return String.Concat(ImplicitPrefix, source);
+		}
+		internal static bool IsImplicit(string source)
+		{
+			return (source.StartsWith(ImplicitPrefix, StringComparison.Ordinal));
```
Comment (paymicro) [inline Xamarin.Forms.Core/Routing.cs]:
```suggestion
			return source.StartsWith(ImplicitPrefix, StringComparison.Ordinal);
```

File: Xamarin.Forms.Core/Routing.cs [hunk 6]
```diff
@@ -11,35 +11,55 @@ public static class Routing
...
+		}
+		internal static bool IsImplicit(Element source)
+		{
+			return IsImplicit(GetRoute(source));
 		}
 
 		internal static bool CompareWithRegisteredRoutes(string compare) => s_routes.ContainsKey(compare);
 
-		internal static bool CompareRoutes(string route, string compare, out bool isImplicit)
+		internal static bool CompareRoutes(string route, string compare)
```
Comment (paymicro) [inline Xamarin.Forms.Core/Routing.cs]: Now this method is not used anywhere

File: Xamarin.Forms.Core/Routing.cs [hunk 7]
```diff
@@ -11,35 +11,55 @@ public static class Routing
...
+		{
+			string[] keys = new string[s_routes.Count];
+			int i = 0;
+			foreach(var key in s_routes.Keys)
+			{
+				keys[i] = key;
+				i++;
+			}
+
+			return keys;
```
Comment (paymicro) [inline Xamarin.Forms.Core/Routing.cs]: a little bit easier
```C#
string[] keys = new string[s_routes.Count];
s_routes.Keys.CopyTo(keys, 0);
return keys;
```

**Comment by PureWeen** [comment]
@samhouts

If it's alright my plan is to sweep through and get rid of everything using Linq for the next Pre, store a few more things locally, etc..

I'd apply a few now but I don't want the UI tests to quit out so we can get this merged and tested

**Pull request merged**